### ISSUE TITLE
Firebase integration

### ISSUE THREAD

**Issue by podviaznikov** [hashobject/quizzy#27]
1. Store users and points in firebase instead of memory.
2. Display leaderboard directly from firebase (without taking data form node.js server)

**Comment by PodviaznikovPavlo** [comment]
okay, its clear

**Comment by PodviaznikovPavlo** [comment]
@podviaznikov 

i have question here. i dont understand '2' task. i am going to do in that way: in nodejs server i connect firebase, get from it data and send to front. So that all data will be on back. And front just get processed data (i`m making sorting on back). 

and you said that i should display leaders directly from firebase, is it mean that i should connect firebase to front? or should i use first variant, that i described above?

**Comment by PodviaznikovPavlo** [comment]
@podviaznikov 

here i have problem with async callbacks. 

what i want to get. 

function getDataFromFB(){
    ref.on('value', function(snap){
        return snap.val();
    })
}

users = getDataFromFB();

or users can be array and in getDataFromFB instead of return i`ll do users = snap.val();

But users are empty, because of callback. I know that in jQuery there is Deffered. 
In node js i found such things 'async' and 'promises'. But i dont understand how to use them. Can you help me with that?

**Comment by podviaznikov** [comment]
> and you said that i should display leaders directly from firebase, is it mean that i should connect firebase to front? or should i use first variant, that i described above?

You should connect firebase directly to the frontend when you want to read data.

**Comment by PodviaznikovPavlo** [comment]
@podviaznikov 

okay, i understand

**Comment by podviaznikov** [comment]
Regarding callbacks. There are different ways to do things but I do them in such way:

```
function getDataFromFB(callback){
  ref.on('value', function(snap){
   callback(snap.val());
  });
}
```

What happens here is instead of `return` you use callback.

Than when you need to get data from firebase you did:

```
var users = getDataFromFB();
```

But instead you need to do :

```
getDataFromFB(function(users){
   console.log('Users are here', users);
});
```

**Comment by PodviaznikovPavlo** [comment]
@podviaznikov 

yes, i try it today and it works. but it has negative side. i need to work inside of this

getDataFromFB(function(users){
   console.log('Users are here', users);
});

I think that it is not a good way, because than i will add some info to FB, and for that issue i need one more callback and so on. There can be too many callbacks. 

This method works, i try it.

But i want to get something that make work that expression  var users = getDataFromFB();
So i want to wait while data will come and then put that data to users and use users in code.

I read about async, wait for, promises. But i didnt get how to work with them

**Comment by podviaznikov** [comment]
In JavaScript you pretty much can't do this.

```
var users = getDataFromFB();
```

You are already using callbacks everywhere (https://github.com/hashobject/quizzy/blob/master/public/main.js#L41, https://github.com/hashobject/quizzy/blob/master/public/main.js#L61, https://github.com/hashobject/quizzy/blob/master/public/main.js#L74)
So I don't see why are you confused in this case. It is exactly the same code as you have before.
